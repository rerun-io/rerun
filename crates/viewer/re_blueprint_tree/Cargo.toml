[package]
authors.workspace = true
description = "The UI for the blueprint tree in the left panel."
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "re_blueprint_tree"
publish = true
readme = "README.md"
repository.workspace = true
rust-version.workspace = true
version.workspace = true
include.workspace = true

[lints]
workspace = true

[package.metadata.docs.rs]
all-features = true

[features]
default = []
testing = [
  "dep:serde",
  "smallvec/serde",
  "re_viewer_context/testing",
  "re_viewport_blueprint/testing",
]


[dependencies]
re_context_menu.workspace = true
re_data_ui.workspace = true
re_entity_db.workspace = true
re_log_types.workspace = true
re_log.workspace = true
re_tracing.workspace = true
re_types.workspace = true
re_ui.workspace = true
re_viewer_context.workspace = true
re_viewport_blueprint.workspace = true

egui.workspace = true
egui_tiles.workspace = true
itertools.workspace = true
serde = { workspace = true, optional = true }
smallvec.workspace = true


[dev-dependencies]
re_chunk_store.workspace = true
re_view_spatial.workspace = true
re_viewer_context = { workspace = true, features = ["testing"] }
re_viewport_blueprint = { workspace = true, features = ["testing"] }

egui_kittest.workspace = true
insta = { workspace = true, features = ["redactions", "yaml"] }
smallvec = { workspace = true, features = ["serde"] }
