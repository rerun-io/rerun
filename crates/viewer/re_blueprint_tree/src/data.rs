//! Data structure describing the contents of the blueprint tree
//!
//! The design goal of these structures is to cover the entire underlying tree (container hierarchy
//! and data result hierarchies) by walking it once entirely, applying filtering on the way.
//!
//! This is done regardless of whether the data is actually used by the UI (e.g. if everything is
//! collapsed). Benchmarks have indicated that this approach incurs a negligible overhead compared
//! to the overall cost of having large blueprint trees (a.k.a the many-entities performance
//! issues: <https://github.com/rerun-io/rerun/issues/8233>).

use std::ops::{ControlFlow, Range};

use itertools::Itertools;
use smallvec::SmallVec;

use re_entity_db::InstancePath;
use re_log_types::external::re_types_core::ViewClassIdentifier;
use re_log_types::EntityPath;
use re_types::blueprint::components::Visible;
use re_ui::filter_widget::{FilterMatcher, PathRanges};
use re_viewer_context::{
    CollapseScope, ContainerId, Contents, ContentsName, DataQueryResult, DataResultNode, Item,
    ViewId, ViewerContext, VisitorControlFlow,
};
use re_viewport_blueprint::{ContainerBlueprint, ViewBlueprint, ViewportBlueprint};

use crate::data_result_node_or_path::DataResultNodeOrPath;

/// Top-level blueprint tree structure.
#[derive(Debug, Default)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub struct BlueprintTreeData {
    pub root_container: Option<ContainerData>,
}

impl BlueprintTreeData {
    pub fn from_blueprint_and_filter(
        ctx: &ViewerContext<'_>,
        viewport_blueprint: &ViewportBlueprint,
        filter_matcher: &FilterMatcher,
    ) -> Self {
        re_tracing::profile_function!();

        Self {
            root_container: viewport_blueprint
                .container(&viewport_blueprint.root_container)
                .and_then(|container_blueprint| {
                    ContainerData::from_blueprint_and_filter(
                        ctx,
                        viewport_blueprint,
                        container_blueprint,
                        filter_matcher,
                    )
                }),
        }
    }

    pub fn visit<B>(
        &self,
        mut visitor: impl FnMut(BlueprintTreeItem<'_>) -> VisitorControlFlow<B>,
    ) -> ControlFlow<B> {
        if let Some(root_container) = &self.root_container {
            root_container.visit(&mut visitor)
        } else {
            ControlFlow::Continue(())
        }
    }
}

// ---

/// Data for either a container or a view (both of which possible child of a container).
#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub enum ContentsData {
    Container(ContainerData),
    View(ViewData),
}

impl ContentsData {
    pub fn visit<B>(
        &self,
        visitor: &mut impl FnMut(BlueprintTreeItem<'_>) -> VisitorControlFlow<B>,
    ) -> ControlFlow<B> {
        match &self {
            Self::Container(container_data) => container_data.visit(visitor),
            Self::View(view_data) => view_data.visit(visitor),
        }
    }
}

/// Data related to a single container and its children.
#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub struct ContainerData {
    #[cfg_attr(feature = "testing", serde(skip))]
    pub id: ContainerId,
    pub name: ContentsName,
    pub kind: egui_tiles::ContainerKind,
    pub visible: bool,
    pub default_open: bool,

    pub children: Vec<ContentsData>,
}

impl ContainerData {
    pub fn from_blueprint_and_filter(
        ctx: &ViewerContext<'_>,
        viewport_blueprint: &ViewportBlueprint,
        container_blueprint: &ContainerBlueprint,
        filter_matcher: &FilterMatcher,
    ) -> Option<Self> {
        let children = container_blueprint
            .contents
            .iter()
            .filter_map(|content| match content {
                Contents::Container(container_id) => {
                    if let Some(container_blueprint) = viewport_blueprint.container(container_id) {
                        Self::from_blueprint_and_filter(
                            ctx,
                            viewport_blueprint,
                            container_blueprint,
                            filter_matcher,
                        )
                        .map(ContentsData::Container)
                    } else {
                        re_log::warn_once!(
                            "Failed to find container {container_id} in ViewportBlueprint"
                        );
                        None
                    }
                }
                Contents::View(view_id) => {
                    if let Some(view_blueprint) = viewport_blueprint.view(view_id) {
                        ViewData::from_blueprint_and_filter(ctx, view_blueprint, filter_matcher)
                            .map(ContentsData::View)
                    } else {
                        re_log::warn_once!("Failed to find view {view_id} in ViewportBlueprint");
                        None
                    }
                }
            })
            .collect_vec();

        // everything was filtered out
        if filter_matcher.is_active() && children.is_empty() {
            return None;
        }

        Some(Self {
            id: container_blueprint.id,
            name: container_blueprint.display_name_or_default(),
            kind: container_blueprint.container_kind,
            visible: container_blueprint.visible,
            default_open: true,
            children,
        })
    }

    pub fn visit<B>(
        &self,
        visitor: &mut impl FnMut(BlueprintTreeItem<'_>) -> VisitorControlFlow<B>,
    ) -> ControlFlow<B> {
        if visitor(BlueprintTreeItem::Container(self)).visit_children()? {
            for child in &self.children {
                child.visit(visitor)?;
            }
        }

        ControlFlow::Continue(())
    }

    pub fn item(&self) -> Item {
        Item::Container(self.id)
    }
}

// ---

/// Data related to a single view and its content.
#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub struct ViewData {
    #[cfg_attr(feature = "testing", serde(skip))]
    pub id: ViewId,

    pub class_identifier: ViewClassIdentifier,

    pub name: ContentsName,
    pub visible: bool,
    pub default_open: bool,

    /// The origin tree contains the data results contained in the subtree defined by the view
    /// origin. They are presented first in the blueprint tree.
    pub origin_tree: Option<DataResultData>,

    /// Projection trees are the various trees contained data results which are outside the view
    /// origin subtrees. They are presented after a "Projections:" label in the blueprint tree.
    ///
    /// These trees may be super-trees of the view origin trees. In that case, the view origin
    /// subtree is represented by a stub item, see [`DataResultKind::OriginProjectionPlaceholder`].
    pub projection_trees: Vec<DataResultData>,
}

impl ViewData {
    fn from_blueprint_and_filter(
        ctx: &ViewerContext<'_>,
        view_blueprint: &ViewBlueprint,
        filter_matcher: &FilterMatcher,
    ) -> Option<Self> {
        re_tracing::profile_function!();

        let query_result = ctx.lookup_query_result(view_blueprint.id);
        let result_tree = &query_result.tree;

        //
        // Data results within the view origin subtree
        //

        let mut hierarchy = Vec::with_capacity(10);
        let mut hierarchy_highlights = PathRanges::default();
        let origin_tree = DataResultData::from_data_result_and_filter(
            ctx,
            view_blueprint,
            query_result,
            &DataResultNodeOrPath::from_path_lookup(result_tree, &view_blueprint.space_origin),
            false,
            &mut hierarchy,
            &mut hierarchy_highlights,
            filter_matcher,
        );

        debug_assert!(hierarchy.is_empty());
        hierarchy.clear();
        hierarchy_highlights.clear();

        //
        // Data results outside the view origin subtree (a.k.a projections)
        //

        let mut projections = Vec::new();
        result_tree.visit(&mut |node| {
            if node
                .data_result
                .entity_path
                .starts_with(&view_blueprint.space_origin)
            {
                // If it's under the origin, we're not interested, stop recursing.
                false
            } else if node.data_result.tree_prefix_only {
                // Keep recursing until we find a projection.
                true
            } else {
                projections.push(node);

                // No further recursion needed in this branch, everything below is included
                // in the projection (or shouldn't be included if the projection has already
                // been filtered out).
                false
            }
        });

        let projection_trees = projections
            .into_iter()
            .filter_map(|node| {
                let projection_tree = DataResultData::from_data_result_and_filter(
                    ctx,
                    view_blueprint,
                    query_result,
                    &DataResultNodeOrPath::DataResultNode(node),
                    true,
                    &mut hierarchy,
                    &mut hierarchy_highlights,
                    filter_matcher,
                );

                debug_assert!(hierarchy.is_empty());
                hierarchy.clear();
                hierarchy_highlights.clear();

                projection_tree
            })
            .collect_vec();

        if origin_tree.is_none() && projection_trees.is_empty() {
            return None;
        }

        let default_open = filter_matcher.is_active()
            || origin_tree.as_ref().map_or(true, |data_result_data| {
                default_open_for_data_result(data_result_data.children.len())
            });

        Some(Self {
            id: view_blueprint.id,
            class_identifier: view_blueprint.class_identifier(),
            name: view_blueprint.display_name_or_default(),
            visible: view_blueprint.visible,
            default_open,
            origin_tree,
            projection_trees,
        })
    }

    pub fn visit<B>(
        &self,
        visitor: &mut impl FnMut(BlueprintTreeItem<'_>) -> VisitorControlFlow<B>,
    ) -> ControlFlow<B> {
        if visitor(BlueprintTreeItem::View(self)).visit_children()? {
            if let Some(origin_tree) = &self.origin_tree {
                origin_tree.visit(visitor)?;
            }

            for projection_tree in &self.projection_trees {
                projection_tree.visit(visitor)?;
            }
        }

        ControlFlow::Continue(())
    }

    pub fn item(&self) -> Item {
        Item::View(self.id)
    }
}

// ---

/// The various kind of things that may be represented in a data result tree.
#[derive(Debug, PartialEq, Eq)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub enum DataResultKind {
    /// This is a regular entity part of a data result (or the tree that contains it).
    EntityPart,

    /// When the view has no data result contained in the view origin subtree, we still display the
    /// origin with a warning styling to highlight what is likely an undesired view configuration.
    EmptyOriginPlaceholder,

    /// Since the view origin tree is displayed on its own, we don't repeat it within projections
    /// and instead replace it with this placeholder.
    OriginProjectionPlaceholder,
}

#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub struct DataResultData {
    pub kind: DataResultKind,
    pub entity_path: EntityPath,
    pub visible: bool,

    #[cfg_attr(feature = "testing", serde(skip))]
    pub view_id: ViewId,

    /// Label that should be used for display.
    ///
    /// This typically corresponds to `entity_path.last().ui_string()` but covers corner cases.
    pub label: String,

    /// The sections within the label that correspond to a filter match and should thus be
    /// highlighted.
    pub highlight_sections: SmallVec<[Range<usize>; 1]>,

    pub default_open: bool,
    pub children: Vec<DataResultData>,
}

impl DataResultData {
    #[allow(clippy::too_many_arguments)]
    fn from_data_result_and_filter(
        ctx: &ViewerContext<'_>,
        view_blueprint: &ViewBlueprint,
        query_result: &DataQueryResult,
        data_result_or_path: &DataResultNodeOrPath<'_>,
        projection: bool,
        hierarchy: &mut Vec<String>,
        hierarchy_highlights: &mut PathRanges,
        filter_matcher: &FilterMatcher,
    ) -> Option<Self> {
        re_tracing::profile_function!();

        // Early out.
        if filter_matcher.matches_nothing() {
            return None;
        }

        let entity_path = data_result_or_path.path().clone();
        let data_result_node = data_result_or_path.data_result_node();
        let visible = data_result_node.is_some_and(|node| node.data_result.is_visible(ctx));

        let entity_part_ui_string = entity_path
            .last()
            .map(|entity_part| entity_part.ui_string());

        let (label, should_pop) = if let Some(entity_part_ui_string) = entity_part_ui_string.clone()
        {
            hierarchy.push(entity_part_ui_string.clone());
            (entity_part_ui_string, true)
        } else {
            ("/ (root)".to_owned(), false)
        };

        //
        // Gather some info about the current node…
        //

        enum LeafOrNot<'a> {
            Leaf,
            NotLeaf(&'a DataResultNode),
        }

        /// Temporary structure to hold local information.
        struct NodeInfo<'a> {
            leaf_or_not: LeafOrNot<'a>,
            kind: DataResultKind,
            default_open: bool,
        }

        #[expect(clippy::manual_map)]
        let node_info = if projection {
            if entity_path == view_blueprint.space_origin {
                Some(NodeInfo {
                    leaf_or_not: LeafOrNot::Leaf,
                    kind: DataResultKind::OriginProjectionPlaceholder,
                    // not hierarchical anyway
                    default_open: false,
                })
            } else if let Some(data_result_node) = data_result_node {
                Some(NodeInfo {
                    leaf_or_not: if data_result_node.children.is_empty() {
                        LeafOrNot::Leaf
                    } else {
                        LeafOrNot::NotLeaf(data_result_node)
                    },
                    kind: DataResultKind::EntityPart,
                    // projections are collapsed by default
                    default_open: filter_matcher.is_active(),
                })
            } else {
                // TODO(ab): what are the circumstances for this? Should we warn about it?
                None
            }
        } else {
            // empty origin case
            if entity_path == view_blueprint.space_origin && data_result_node.is_none() {
                Some(NodeInfo {
                    leaf_or_not: LeafOrNot::Leaf,
                    kind: DataResultKind::EmptyOriginPlaceholder,
                    // not hierarchical anyway
                    default_open: false,
                })
            } else if let Some(data_result_node) = data_result_node {
                Some(NodeInfo {
                    leaf_or_not: if data_result_node.children.is_empty() {
                        LeafOrNot::Leaf
                    } else {
                        LeafOrNot::NotLeaf(data_result_node)
                    },
                    kind: DataResultKind::EntityPart,
                    default_open: filter_matcher.is_active()
                        || default_open_for_data_result(data_result_node.children.len()),
                })
            } else {
                // TODO(ab): what are the circumstances for this? Should we warn about it?
                None
            }
        };

        //
        // …then handle the node accordingly.
        //

        let result = node_info.and_then(|node_info| {
            let (is_this_a_match, children) = match node_info.leaf_or_not {
                LeafOrNot::Leaf => {
                    // Key insight: we only ever need to match the hierarchy from the leaf nodes.
                    // Non-leaf nodes know they are a match if any child remains after walking their
                    // subtree.

                    let highlights =
                        filter_matcher.match_path(hierarchy.iter().map(String::as_str));

                    let is_this_a_match = if let Some(highlights) = highlights {
                        hierarchy_highlights.merge(highlights);
                        true
                    } else {
                        false
                    };

                    (is_this_a_match, vec![])
                }

                LeafOrNot::NotLeaf(data_result_node) => {
                    let mut children = data_result_node
                        .children
                        .iter()
                        .filter_map(|child_handle| {
                            let child_node = query_result.tree.lookup_node(*child_handle);

                            debug_assert!(
                                child_node.is_some(),
                                "DataResultNode {data_result_node:?} has an invalid child"
                            );

                            child_node.and_then(|child_node| {
                                Self::from_data_result_and_filter(
                                    ctx,
                                    view_blueprint,
                                    query_result,
                                    &DataResultNodeOrPath::DataResultNode(child_node),
                                    projection,
                                    hierarchy,
                                    hierarchy_highlights,
                                    filter_matcher,
                                )
                            })
                        })
                        .collect_vec();

                    // This is needed because `DataResultNode` stores children in a `SmallVec`, offering
                    // no guarantees about ordering.
                    children.sort_by(|a, b| a.entity_path.cmp(&b.entity_path));

                    let is_this_a_match = !children.is_empty();

                    (is_this_a_match, children)
                }
            };

            is_this_a_match.then(|| {
                let highlight_sections =
                    hierarchy_highlights.remove(hierarchy.len().saturating_sub(1));

                // never highlight the placeholder
                let highlight_sections =
                    if node_info.kind == DataResultKind::OriginProjectionPlaceholder {
                        SmallVec::new()
                    } else {
                        highlight_sections
                            .map(Iterator::collect)
                            .unwrap_or_default()
                    };

                Self {
                    kind: node_info.kind,
                    entity_path,
                    visible,
                    view_id: view_blueprint.id,
                    label,
                    highlight_sections,
                    default_open: node_info.default_open,
                    children,
                }
            })
        });

        if should_pop {
            hierarchy_highlights.remove(hierarchy.len().saturating_sub(1));
            hierarchy.pop();
        }

        result
    }

    pub fn visit<B>(
        &self,
        visitor: &mut impl FnMut(BlueprintTreeItem<'_>) -> VisitorControlFlow<B>,
    ) -> ControlFlow<B> {
        if visitor(BlueprintTreeItem::DataResult(self)).visit_children()? {
            for child in &self.children {
                child.visit(visitor)?;
            }
        }

        ControlFlow::Continue(())
    }

    pub fn item(&self) -> Item {
        Item::DataResult(self.view_id, self.instance_path())
    }

    pub fn instance_path(&self) -> InstancePath {
        self.entity_path.clone().into()
    }

    /// Update the visibility of this data result.
    pub fn update_visibility(&self, ctx: &ViewerContext<'_>, visible: bool) {
        let query_result = ctx.lookup_query_result(self.view_id);
        let result_tree = &query_result.tree;
        if let Some(data_result) = result_tree.lookup_result_by_path(&self.entity_path) {
            data_result.save_recursive_override_or_clear_if_redundant(
                ctx,
                &query_result.tree,
                &Visible::from(visible),
            );
        }
    }

    /// Remove this data result from the view.
    pub fn remove_data_result_from_view(
        &self,
        ctx: &ViewerContext<'_>,
        viewport_blueprint: &ViewportBlueprint,
    ) {
        if let Some(view_blueprint) = viewport_blueprint.view(&self.view_id) {
            view_blueprint
                .contents
                .remove_subtree_and_matching_rules(ctx, self.entity_path.clone());
        }
    }
}

/// If a group or view has a total of this number of elements, show its subtree by default?
//TODO(ab): taken from legacy implementation, does it still make sense?
fn default_open_for_data_result(num_children: usize) -> bool {
    2 <= num_children && num_children <= 3
}

// ---

/// Wrapper structure used for the closure of [`BlueprintTreeData::visit`] and friends.
#[derive(Debug, Clone, Copy)]
pub enum BlueprintTreeItem<'a> {
    Container(&'a ContainerData),
    View(&'a ViewData),
    DataResult(&'a DataResultData),
}

impl BlueprintTreeItem<'_> {
    pub fn item(&self) -> Item {
        match self {
            BlueprintTreeItem::Container(container_data) => container_data.item(),
            BlueprintTreeItem::View(view_data) => view_data.item(),
            BlueprintTreeItem::DataResult(data_result_data) => data_result_data.item(),
        }
    }

    pub fn default_open(&self) -> bool {
        match self {
            BlueprintTreeItem::Container(container_data) => container_data.default_open,
            BlueprintTreeItem::View(view_data) => view_data.default_open,
            BlueprintTreeItem::DataResult(data_result_data) => data_result_data.default_open,
        }
    }

    pub fn is_open(&self, ctx: &egui::Context, collapse_scope: CollapseScope) -> Option<bool> {
        collapse_scope.item(self.item()).map(|collapse_id| {
            collapse_id
                .is_open(ctx)
                .unwrap_or_else(|| self.default_open())
        })
    }
}
