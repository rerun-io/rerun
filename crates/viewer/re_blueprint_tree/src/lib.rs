//! This crate implements the UI for the blueprint tree in the left panel.

mod blueprint_tree;

pub use blueprint_tree::BlueprintTree;
