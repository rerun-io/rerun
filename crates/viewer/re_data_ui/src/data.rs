use re_viewer_context::{UiLayout, ViewerContext};

use super::DataUi;

impl DataUi for re_types::datatypes::Mat3x3 {
    fn data_ui(
        &self,
        _ctx: &ViewerContext<'_>,
        ui: &mut egui::Ui,
        _ui_layout: UiLayout,
        _query: &re_chunk_store::LatestAtQuery,
        _db: &re_entity_db::EntityDb,
    ) {
        egui::Grid::new("mat3").num_columns(3).show(ui, |ui| {
            ui.monospace(self[0].to_string());
            ui.monospace(self[3].to_string());
            ui.monospace(self[6].to_string());
            ui.end_row();

            ui.monospace(self[1].to_string());
            ui.monospace(self[4].to_string());
            ui.monospace(self[7].to_string());
            ui.end_row();

            ui.monospace(self[2].to_string());
            ui.monospace(self[5].to_string());
            ui.monospace(self[8].to_string());
            ui.end_row();
        });
    }
}
