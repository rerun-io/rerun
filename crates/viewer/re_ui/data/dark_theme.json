{
  "Alias": {
    "Typography": {
      "Default": {
        "value": "{Global.Typography.200}",
        "description": "Default font in Rerun's UI",
        "type": "typography"
      }
    }
  },
  "Global": {
    "Typography": {
      "100": {
        "value": {
          "fontSize": "10px",
          "fontWeight": "Medium",
          "fontFamily": "Inter",
          "lineHeight": "12px",
          "letterSpacing": "-0.12px"
        },
        "description": "",
        "type": "typography"
      },
      "200": {
        "value": {
          "fontSize": "12px",
          "fontWeight": "Medium",
          "fontFamily": "Inter",
          "lineHeight": "16px",
          "letterSpacing": "-0.15px"
        },
        "description": "",
        "type": "typography"
      }
    }
  }
}
