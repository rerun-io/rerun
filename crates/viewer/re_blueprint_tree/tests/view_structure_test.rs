//! Snapshot testTest suite dedicated to snapshot the way we present various kinds of blueprint tree structures
//! with a focus on various view contents and filter configuration.

#![cfg(feature = "testing")]

use egui::Vec2;
use egui_kittest::{SnapshotError, SnapshotOptions};
use itertools::Itertools;

use re_blueprint_tree::data::BlueprintTreeData;
use re_blueprint_tree::BlueprintTree;
use re_chunk_store::external::re_chunk::ChunkBuilder;
use re_chunk_store::RowId;
use re_log_types::{build_frame_nr, EntityPath, Timeline};
use re_types::archetypes::Points3D;
use re_ui::filter_widget::FilterState;
use re_viewer_context::test_context::TestContext;
use re_viewer_context::{RecommendedView, ViewClass, ViewId};
use re_viewport_blueprint::test_context_ext::TestContextExt;
use re_viewport_blueprint::{ViewBlueprint, ViewportBlueprint};

const VIEW_ID: &str = "this-is-a-view-id";

#[derive(Debug, Clone, Copy)]
enum RecordingKind {
    /// No entities are logged.
    Empty,

    /// Some placeholder entity hierarchy is logged.
    Regular,
}

struct TestCase {
    name: &'static str,

    origin: EntityPath,
    entity_filter: &'static str,

    recording_kind: RecordingKind,
}

fn base_test_cases() -> impl Iterator<Item = TestCase> {
    [
        TestCase {
            name: "empty",
            origin: EntityPath::root(),
            entity_filter: "$origin/**",
            recording_kind: RecordingKind::Empty,
        },
        TestCase {
            name: "root_origin",
            origin: EntityPath::root(),
            entity_filter: "$origin/**",
            recording_kind: RecordingKind::Regular,
        },
        TestCase {
            name: "non_root_origin",
            origin: EntityPath::from("/path/to"),
            entity_filter: "$origin/**",

            recording_kind: RecordingKind::Regular,
        },
        TestCase {
            name: "unknown_origin",
            origin: EntityPath::from("/wrong/path"),
            entity_filter: "$origin/**",
            recording_kind: RecordingKind::Regular,
        },
        TestCase {
            name: "single_proj",
            origin: EntityPath::from("/center/way"),
            entity_filter: "$origin/**\n/path/to/**",
            recording_kind: RecordingKind::Regular,
        },
        TestCase {
            name: "proj_with_placeholder",
            origin: EntityPath::from("/path/to"),
            entity_filter: "/**",
            recording_kind: RecordingKind::Regular,
        },
        TestCase {
            name: "multiple_proj",
            origin: EntityPath::from("/center/way"),
            entity_filter: "$origin/**\n/path/to/right\n/path/to/the/**",
            recording_kind: RecordingKind::Regular,
        },
    ]
    .into_iter()
}

fn filter_queries() -> impl Iterator<Item = Option<&'static str>> {
    [
        None,
        Some("t"),
        Some("void"),
        Some("path"),
        Some("ath t"),
        Some("ath left"),
        Some("to/the"),
        Some("/to/the/"),
        Some("to/the oid"),
        Some("/path/to /rig"),
        Some("/path/to/th"),
    ]
    .into_iter()
}

fn test_context(test_case: &TestCase) -> TestContext {
    let mut test_context = TestContext::default();

    test_context.register_view_class::<re_view_spatial::SpatialView3D>();

    match test_case.recording_kind {
        RecordingKind::Empty => {}
        RecordingKind::Regular => {
            test_context.log_entity("/path/to/left".into(), add_point_to_chunk_builder);
            test_context.log_entity("/path/to/right".into(), add_point_to_chunk_builder);
            test_context.log_entity("/path/to/the/void".into(), add_point_to_chunk_builder);
            test_context.log_entity("/path/onto/their/coils".into(), add_point_to_chunk_builder);
            test_context.log_entity("/center/way".into(), add_point_to_chunk_builder);
        }
    }

    test_context.setup_viewport_blueprint(|_, blueprint| {
        let view = ViewBlueprint::new_with_id(
            re_view_spatial::SpatialView3D::identifier(),
            RecommendedView {
                origin: test_case.origin.clone(),
                query_filter: test_case
                    .entity_filter
                    .try_into()
                    .expect("invalid entity filter"),
            },
            ViewId::hashed_from_str(VIEW_ID),
        );

        blueprint.add_views(std::iter::once(view), None, None);
    });

    test_context
}

fn add_point_to_chunk_builder(builder: ChunkBuilder) -> ChunkBuilder {
    builder.with_archetype(
        RowId::new(),
        [build_frame_nr(0)],
        &Points3D::new([[0.0, 0.0, 0.0]]),
    )
}

#[test]
fn test_all_snapshot_test_cases() {
    let errors = filter_queries()
        .flat_map(|filter_query| {
            base_test_cases()
                .map(move |test_case| (filter_query, run_test_case(&test_case, filter_query)))
        })
        .filter_map(|(filter_query, result)| result.err().map(|err| (filter_query, err)))
        .collect_vec();

    for (filter_query, error) in &errors {
        eprintln!("ERR: filter '{filter_query:?}': {error:?}");
    }

    assert!(errors.is_empty(), "Some test cases failed");
}

fn run_test_case(test_case: &TestCase, filter_query: Option<&str>) -> Result<(), SnapshotError> {
    let mut test_context = test_context(test_case);
    let view_id = ViewId::hashed_from_str(VIEW_ID);

    let mut blueprint_tree = BlueprintTree::default();

    // This trick here is to run the blueprint panel for a frame, such that it registers the current
    // application id. This way, the blueprint panel will not discard the filter state we set up
    // when it's run for the snapshot.
    test_context.run_in_egui_central_panel(|ctx, ui| {
        let blueprint =
            ViewportBlueprint::try_from_db(ctx.store_context.blueprint, ctx.blueprint_query);

        blueprint_tree.show(ctx, &blueprint, ui);
    });

    if let Some(filter_query) = filter_query {
        blueprint_tree.activate_filter(filter_query);
    }

    // set the current timeline to the timeline where data was logged to
    test_context.set_active_timeline(Timeline::new_sequence("frame_nr"));

    let mut harness = test_context
        .setup_kittest_for_rendering()
        .with_size(Vec2::new(400.0, 800.0))
        .build_ui(|ui| {
            test_context.run(&ui.ctx().clone(), |viewer_ctx| {
                re_context_menu::collapse_expand::collapse_expand_view(
                    viewer_ctx,
                    &view_id,
                    blueprint_tree.collapse_scope(),
                    true,
                );

                let blueprint = ViewportBlueprint::try_from_db(
                    viewer_ctx.store_context.blueprint,
                    viewer_ctx.blueprint_query,
                );

                blueprint_tree.show(viewer_ctx, &blueprint, ui);
            });

            test_context.handle_system_commands();
        });

    harness.run();

    let options = SnapshotOptions::default().output_path(format!(
        "tests/snapshots/view_structure_test/{}",
        filter_query
            .map(|query| format!("query-{}", query.replace(' ', ",").replace('/', "_")))
            .unwrap_or("no-query".to_owned())
    ));
    harness.try_snapshot_options(test_case.name, &options)
}

// ---

#[test]
fn test_all_insta_test_cases() {
    for test_case in base_test_cases() {
        for filter_query in filter_queries() {
            let test_context = test_context(&test_case);

            let blueprint_tree_data =
                test_context.run_once_in_egui_central_panel(|viewer_ctx, _| {
                    let blueprint = ViewportBlueprint::try_from_db(
                        viewer_ctx.store_context.blueprint,
                        viewer_ctx.blueprint_query,
                    );

                    let mut filter_state = FilterState::default();

                    if let Some(filter_query) = filter_query {
                        filter_state.activate(filter_query);
                    }

                    BlueprintTreeData::from_blueprint_and_filter(
                        viewer_ctx,
                        &blueprint,
                        &filter_state.filter(),
                    )
                });

            let mut settings = insta::Settings::clone_current();
            settings.set_prepend_module_to_snapshot(false);
            settings.set_snapshot_path(format!(
                "snapshots/view_structure_test/{}",
                filter_query
                    .map(|query| format!("query-{}", query.replace(' ', ",").replace('/', "_")))
                    .unwrap_or("no-query".to_owned())
            ));

            settings.bind(|| {
                insta::assert_yaml_snapshot!(test_case.name, blueprint_tree_data);
            });
        }
    }
}
