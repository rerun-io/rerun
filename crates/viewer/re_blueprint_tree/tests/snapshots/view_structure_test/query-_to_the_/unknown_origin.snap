---
source: crates/viewer/re_blueprint_tree/tests/view_structure_test.rs
expression: blueprint_tree_data
snapshot_kind: text
---
root_container: ~
